lib/min_heap.rb

```diff
@@ -44,28 +51,52 @@ def to_s
   end
 
   # This method returns true if the heap is empty
-  # Time complexity: ?
-  # Space complexity: ?
+  # Time complexity: o(1)
+  # Space complexity: o(1)
   def empty?
-    raise NotImplementedError, "Method not implemented yet..."
+    return @store.length == 0 ? true : false
   end
 
   private
 
   # This helper method takes an index and
   #  moves it up the heap, if it is less than it's parent node.
   #  It could be **very** helpful for the add method.
-  # Time complexity: ?
-  # Space complexity: ?
+  # Time complexity: o(log n)
+  # Space complexity: o(1)
   def heap_up(index)
-    
+    # base cases
+    return if index == 0
+    return if @store[(index-1)/2].key < @store[index].key
+
+    # if the parent is bigger then itself
+    if @store[(index-1)/2].key > @store[index].key
+      swap((index-1)/2, index)
+      heap_up((index-1)/2)
+    end
+    return 
   end
 
   # This helper method takes an index and 
   #  moves it up the heap if it's smaller
   #  than it's parent node.
   def heap_down(index)
-    raise NotImplementedError, "Method not implemented yet..."
+    left_i = 2*index+1
+    right_i = 2*index+2
+    # base case
+    # if there is no more children
+    return if left_i > @store.length-1 
+
+    # find the min_child
+    min_i = left_i
+    if right_i < @store.length
```

Comment:
Nice good catch checking to see if the right index is past the end of the array.